### PR TITLE
Move DNS records from base_domain to cluster_domain

### PR DESCRIPTION
**Pull request by abhinavdahiya**
The issue was reported in https://github.com/openshift/installer/issues/1136

On AWS, we are currently creating a private zone for `base_domain` and create all the necessary records in that private zone. when users create a new cluster under the same domain, we create a new private zone for the `base_domain` again. This setup creates 2 zones each with authority on `base_domain` and therefore each cluster cannot resolve the api or other endpoints for the other cluster.

A solution is that we create private zones for a cluster with authority on `cluster_domain = cluster_name.base_domain`. This allows each cluster to maintain authority on subdomain of basedomain and allows each to resolve the other.

some of our current dns records look like:
`cluster_name-api.base_domain` -> for api and ignition server.
`cluster_name-etcd-{idx}.base_domain` -> for each master with etcd.

To make sure when moving records from base_domain to cluster_domain, we do not make our dns names very long. the new records look like:
`api.cluster_domain` -> for api and ignition server. 
`etcd-{idx}.cluster_domain` -> for each master with etcd.
This keeps all the records exactly the same length as before, as we are moving cluster_name and replacing `-` with a `.`

tasks:

- [x] move libvirt records from `base_domain` to be under `cluster_domain` with new schema
- [x] move aws records from `base_domain` to be under `cluster_domain` with new schema
- [x] move openstack records from `base_domain` to be under `cluster_domain` with new schema
- [x] changes are required in Machine Config Operator to use the new api server url
- [x] changes are required in Ingress Operator to make sure it can create the `*.app.cluster_domain`
- [ ] possible changes required in [`DNS.config.openshift.io`](https://github.com/openshift/api/blob/4912e102a00fa8bb4bf27f83808dce6ef1b6887d/config/v1/types_dns.go#L22-L29) and consumers.

### CODE REVIEW

File: data/data/aws/master/variables.tf [hunk 1]
```diff
@@ -1,8 +1,3 @@
-variable "base_domain" {
```
Comment (wking) [inline data/data/aws/master/variables.tf]: Oops, looks like I added this without a consumer in 0bec46d3 (#362).  Can you pull it out into a separate PR so we can land it without waiting for this PR to settle?

Comment (abhinavdahiya) [reply data/data/aws/master/variables.tf]: sure.

File: data/data/aws/route53/base.tf [hunk 2]
```diff
@@ -6,11 +6,13 @@ locals {
   public_zone_id = "${data.aws_route53_zone.base.zone_id}"
 
   zone_id = "${var.private_zone_id}"
+
+  cluster_domain = "${var.cluster_name}.${var.base_domain}"
```
Comment (wking) [inline data/data/aws/route53/base.tf]: This package could be adjusted to take the unsplit cluster domain if we replaced [this][1] with Go logic to snip subdomains off the full cluster name until we found a public zone.  Or we could restructure the install-config to make the public zone's domain part of the AWS-specific platform configuration.  Both of those are probably more work than we want to sink into initial exploration, but I thought I'd file a note to track future polish possibilities ;).

[1]: https://github.com/openshift/installer/blob/v0.11.0/data/data/aws/route53/base.tf#L1-L3

File: data/data/aws/vpc/vpc.tf [hunk 3]
```diff
@@ -1,6 +1,8 @@
 locals {
   new_worker_cidr_range = "${cidrsubnet(data.aws_vpc.cluster_vpc.cidr_block,1,1)}"
   new_master_cidr_range = "${cidrsubnet(data.aws_vpc.cluster_vpc.cidr_block,1,0)}"
+
+  cluster_domain = "${var.cluster_name}.${var.base_domain}"
```
Comment (wking) [inline data/data/aws/vpc/vpc.tf]: This module could be adjusted to take the unsplit cluster domain and use it where it currently uses `var.cluster_name` in tag values and resource names.  That would address #762, although we might have to use the length-limited cluster ID for the load-balancer names or somehow introduce uniqueness there.  I've just filed #1170 making cluster-names in `metadata.json` libvirt-specific, so this resource naming is the last #762 blocker.

**Comment by abhinavdahiya** [comment]
> changes are required in Machine Config Operator to use the new api server url
I have some PRs in flight, https://github.com/openshift/installer/pull/1155 and https://github.com/openshift/machine-config-operator/pull/357

File: data/data/aws/vpc/vpc.tf (hunk 3 above)
Comment (abhinavdahiya) [reply data/data/aws/vpc/vpc.tf]: https://github.com/openshift/installer/issues/762 @blrm was looking at finding the resources that would conflict if we try to use same cluster name for same base domain. I remember him mentioning that it was more than this module.

@blrm do you have some notes/finding that can provide more details on what might collide in #762  :innocent:

Comment (wking) [reply data/data/aws/vpc/vpc.tf]: > @blrm was looking at finding the resources that would conflict if we try to use same cluster name for same base domain.

I'd guess:

```console
$ git describe
unreleased-master-177-g4907cba
$ git grep ' name *= .*cluster_name' data/data/aws | cat
data/data/aws/bootstrap/main.tf:  name = "${var.cluster_name}-bootstrap-profile"
data/data/aws/bootstrap/main.tf:  name  = "${var.cluster_name}-bootstrap-role"
data/data/aws/bootstrap/main.tf:  name  = "${var.cluster_name}-bootstrap-policy"
data/data/aws/iam/main.tf:  name = "${var.cluster_name}-worker-profile"
data/data/aws/iam/main.tf:  name  = "${var.cluster_name}-worker-role"
data/data/aws/iam/main.tf:  name  = "${var.cluster_name}_worker_policy"
data/data/aws/main.tf:  name    = "${var.cluster_name}-etcd-${count.index}"
data/data/aws/main.tf:  name    = "_etcd-server-ssl._tcp.${var.cluster_name}"
data/data/aws/master/main.tf:  name = "${var.cluster_name}-master-profile"
data/data/aws/master/main.tf:  name  = "${var.cluster_name}-master-role"
data/data/aws/master/main.tf:  name  = "${var.cluster_name}_master_policy"
data/data/aws/route53/base.tf:  name    = "${var.cluster_name}-api.${var.base_domain}"
data/data/aws/route53/base.tf:  name    = "${var.cluster_name}-api.${var.base_domain}"
data/data/aws/vpc/master-elb.tf:  name                             = "${var.cluster_name}-int"
data/data/aws/vpc/master-elb.tf:  name                             = "${var.cluster_name}-ext"
data/data/aws/vpc/master-elb.tf:  name     = "${var.cluster_name}-api-int"
data/data/aws/vpc/master-elb.tf:  name     = "${var.cluster_name}-api-ext"
data/data/aws/vpc/master-elb.tf:  name     = "${var.cluster_name}-services"
```

but I don't see a problem with removing those in stages.

**Comment by abhinavdahiya** [comment]
>  changes are required in Ingress Operator to make sure it can create the `*.app.cluster_domain`

PR in flight https://github.com/openshift/cluster-ingress-operator/pull/117

File: pkg/asset/manifests/dns.go [hunk 4]
```diff
@@ -54,7 +54,7 @@ func (d *DNS) Generate(dependencies asset.Parents) error {
 			// not namespaced
 		},
 		Spec: configv1.DNSSpec{
-			BaseDomain: installConfig.Config.BaseDomain,
+			BaseDomain: installConfig.Config.ClusterDomain(),
```
Comment (ironcladlou) [inline pkg/asset/manifests/dns.go]: Do we instead need to change `DNSSpec` like:

```go
type DNSSpec struct {
  BaseDomain string `json:"baseDomain"`
  ClusterDomain string `json:"baseDomain"`
}
```

Otherwise downstream consumers (e.g. the ingress operator's DNS manager) are left to guess the public zone.

Comment (abhinavdahiya) [reply pkg/asset/manifests/dns.go]: > Do we instead need to change `DNSSpec` like:
> 
> ```go
> type DNSSpec struct {
>   BaseDomain string `json:"baseDomain"`
>   ClusterDomain string `json:"clusterDomain"`
> }
> ```

I think the `Basedomain` field correctly represents what is required.
from [doc](https://github.com/openshift/api/blob/d67473e7f1907b74d1f27706260eecf0bc9f2a52/config/v1/types_dns.go#L23-L27)
```
	// baseDomain is the base domain of the cluster. All managed DNS records will
	// be sub-domains of this base.
	//
	// For example, given the base domain `openshift.example.com`, an API server
	// DNS record may be created for `cluster-api.openshift.example.com`.
```

while the Ingress spec only specifies the `domain` used for routes
from [doc](https://github.com/openshift/api/blob/d67473e7f1907b74d1f27706260eecf0bc9f2a52/config/v1/types_ingress.go#L23-L26)
```
	// domain is used to generate a default host name for a route when the
	// route's host name is empty.  The generated host name will follow this
	// pattern: "<route-name>.<route-namespace>.<domain>".
```
but does not specify how/where to realize that information for ingress's DNS manager.

Comment (abhinavdahiya) [reply pkg/asset/manifests/dns.go]: So what if we extend the `DNSSpec` to 
```go
type DNSSpec struct {
  BaseDomain string `json:"baseDomain"`
  PublicZone *DNSZone `json:"publicZone"`
  PrivateZone *DNSZone `json:"privateZone"`
}

// DNSZone describes a dns zone for a provider.
// A zone can be identified by an identifier or tags.
type DNSZone struct {
  // id is the identifier that can be used to find the dns zone
  // +optional
  ID *string `json:"id"`
  // tags is a map of tags that can be used to query the dns zone
  // +optional
  Tags map[string]string `json:"tags"`
}
```

Comment (abhinavdahiya) [reply pkg/asset/manifests/dns.go]: @ironcladlou https://github.com/openshift/api/pull/202

Comment (ironcladlou) [reply pkg/asset/manifests/dns.go]: https://github.com/openshift/cluster-ingress-operator/pull/121 covers step 3.

**Comment by abhinavdahiya** [comment]
> > changes are required in Machine Config Operator to use the new api server url
> 
> I have some PRs in flight, #1155 and [openshift/machine-config-operator#357](https://github.com/openshift/machine-config-operator/pull/357)

merged :yay:

**Comment by abhinavdahiya** [comment]
@flaper87 @russellb 
do you have pointers on how to move openstack to this new scheme?
cluster_domain=cluster_name.base_domain
apiserver `api.cluster_domain`
etcd servers `etcd-$idx.cluster_domain`

see some stuff in `data/data/openstack/service/main.tf`

**Comment by abhinavdahiya** [comment]
needs https://github.com/openshift/cluster-authentication-operator/pull/72 
it was hardcoding the apiserver URL to old structure.

**Comment by wking** [comment]
> needs openshift/cluster-authentication-operator#72

/hold

Are we waiting on that PR?  Or just assuming we'll fail e2e until that PR lands?

**Comment by flaper87** [comment]
I'll test this patch on OpenStack asap

File: data/data/openstack/service/main.tf [hunk 5]
```diff
@@ -168,20 +168,20 @@ data "ignition_file" "coredb" {
...
                                 7200       ; refresh (2 hours)
                                 3600       ; retry (1 hour)
                                 1209600    ; expire (2 weeks)
                                 3600       ; minimum (1 hour)
                                 )
 
-${var.cluster_name}-api  IN  A  ${var.service_vm_floating_ip}
-*.apps.${var.cluster_name}  IN  A  ${var.service_vm_floating_ip}
+api  IN  A  ${var.service_vm_floating_ip}
+*.apps  IN  A  ${var.service_vm_floating_ip}
```
Comment (tomassedovic) [inline data/data/openstack/service/main.tf]: `var.service_vm_floating_ip` is now `var.lb_floating_ip`

Comment (abhinavdahiya) [reply data/data/openstack/service/main.tf]: Thanks! done with https://github.com/openshift/installer/pull/1169/commits/15dfaa2b7c7422b2c7e686b6bdce5895ccc9beb2 -> https://github.com/openshift/installer/pull/1169/commits/6a84e7bc0575f2009765d49229c771e471a9f09f

**Comment by abhinavdahiya** [comment]
> > needs [openshift/cluster-authentication-operator#72](https://github.com/openshift/cluster-authentication-operator/pull/72)
> 
> /hold
> 
> Are we waiting on that PR? Or just assuming we'll fail e2e until that PR lands?

Looks like we will have to wait for [openshift/cluster-authentication-operator#59](https://github.com/openshift/cluster-authentication-operator/pull/59)

**Comment by abhinavdahiya** [comment]
rebased around #1157

File: data/data/openstack/service/variables.tf [hunk 6]
```diff
@@ -19,7 +19,7 @@ variable "cluster_id" {
 
 variable "cluster_domain" {
   type        = "string"
-  description = "The domain name of the cluster."
+  description = "The domain name of the cluster. All DNS recoreds must be under this domain."
```
[the remark this answers is not in the document]
Comment (abhinavdahiya) [reply data/data/openstack/service/variables.tf]: done with https://github.com/openshift/installer/pull/1169/commits/e8ff0892d5e4bc31d8fda961ec7c9bc96628a088 -> https://github.com/openshift/installer/pull/1169/commits/a0a77add3ddf259f7c0dbcfabc00a61af5cc9575

**Comment by crawford** [comment]
https://github.com/openshift/cluster-authentication-operator/pull/72 has been merged.

/retest

**Comment by abhinavdahiya** [comment]
last [e2e-aws](https://openshift-gce-devel.appspot.com/build/origin-ci-test/pr-logs/pull/openshift_installer/1169/pull-ci-openshift-installer-master-e2e-aws/3828) failed due to router flake.
```console
Flaky tests:

[Conformance][Area:Networking][Feature:Router] The HAProxy router should expose prometheus metrics for a route [Suite:openshift/conformance/parallel/minimal]

Failing tests:

[Conformance][Area:Networking][Feature:Router] The HAProxy router should enable openshift-monitoring to pull metrics [Suite:openshift/conformance/parallel/minimal]
```

No more auth errors.
/hold cancel

/retest

**Comment by crawford** [comment]
Known [failures](https://openshift-gce-devel.appspot.com/build/origin-ci-test/pr-logs/pull/openshift_installer/1169/pull-ci-openshift-installer-master-e2e-aws/3828):

```
Flaky tests:

[Conformance][Area:Networking][Feature:Router] The HAProxy router should expose prometheus metrics for a route [Suite:openshift/conformance/parallel/minimal]

Failing tests:

[Conformance][Area:Networking][Feature:Router] The HAProxy router should enable openshift-monitoring to pull metrics [Suite:openshift/conformance/parallel/minimal]
```

/retest
